Cesium3DTilesSelection/src/CesiumIonTilesetLoader.cpp

```diff
@@ -12,6 +12,106 @@
 #include <unordered_map>
 
 namespace Cesium3DTilesSelection {
+
+// This is an IAssetAccessor decorator that handles token refresh for any asset
+// that returns a 401 error.
+class CesiumIonAssetAccessor
+    : public std::enable_shared_from_this<CesiumIonAssetAccessor>,
+      public CesiumAsync::IAssetAccessor {
+public:
+  CesiumIonAssetAccessor(
+      CesiumIonTilesetLoader& tilesetLoader,
+      const std::shared_ptr<CesiumAsync::IAssetAccessor>& pAggregatedAccessor)
+      : _pTilesetLoader(&tilesetLoader),
+        _pAggregatedAccessor(pAggregatedAccessor) {}
+
+  CesiumAsync::Future<std::shared_ptr<CesiumAsync::IAssetRequest>>
+  get(const CesiumAsync::AsyncSystem& asyncSystem,
+      const std::string& url,
+      const std::vector<THeader>& headers = {}) override {
+    // If token refresh is needed, this lambda will be called in the main thread
+    // so that it can safely use the tileset loader.
+    auto refreshToken =
+        [pThis = this->shared_from_this()](
+            const CesiumAsync::AsyncSystem& asyncSystem,
+            std::shared_ptr<CesiumAsync::IAssetRequest>&& pRequest) {
+          if (!pThis->_pTilesetLoader) {
```

Comment:
`CesiumIonAssetAccessor` has a `notifyLoaderIsBeingDestroyed` method which sets it to nullptr. It's called from `CesiumIonTilesetLoader`'s destructor. So yeah, it works like a weak_ptr, except weak_ptr can't be used here because the lifetime of TilesetContentLoaders is not managed by a shared_ptr.

Reply:
My bad for not seeing `notifyLoaderIsBeingDestroyed`. It might be worth worrying -- some day -- about the visibility of the null write with respect to the `CesiumIonTilesetLoader` being destroyed in another thread.

Reply:
`CesiumIonTilesetLoader` is owned by `TilesetContentManager`, which is derived from `ReferenceCountedNonThreadSafe`, which means it is illegal to call `addReference` / `releaseReference` from anywhere other than the thread that owns it (created it). This is checked via assertion in debug builds. So that guarantees that `TilesetContentManager`'s destructor will be called in the main thread, which guarantees `CesiumIonTilesetLoader` will be destroyed in the main thread, too. So `notifyLoaderIsBeingDestroyed` will only ever be called in the main thread.

Other than that, the only thing that ever touches `_pTilesetLoader` is the `refreshToken` lambda, which is also guaranteed to run in the main thread.

Therefore the reads and writes of `_pTilesetLoader` always happen in one thread, so there's no need for synchronization.